### PR TITLE
Refactor to improve the user experience with non-ASCII characters

### PR DESCRIPTION
**Pull request by seisman**
See https://github.com/GenericMappingTools/pygmt/issues/2204#issuecomment-2079661330 for background and the issue.

This PR solves the issue by:

- Generating the mapping dictionary based on the mapping tables in https://github.com/seisman/GMT-octal-codes
- Add the encoding tables to the documentation

**Preview**: 
The tables of supported encodings are at
https://pygmt-dev--3206.org.readthedocs.build/en/3206/techref/encodings.html.

This page is in the "Technical Reference" section, as discussed in #2834.

**Note to maintainers**

Below is the script used to generate the Markdown tables in the documentation
```python
from pygmt.encodings import charset

undefined = "\ufffd"
markdown = ""
for charset_name, mappings in charset.items():
    markdown += f"## Adobe {charset_name} Encoding\n\n"
    markdown += "| octal | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 |\n"
    markdown += "|---|---|---|---|---|---|---|---|---|\n"
    for i in range(0o00, 0o400, 8):
        chars = [mappings.get(j, undefined) for j in range(i, i + 8)]
        if chars == [undefined] * 8:
            continue
        chars = [f"&#x{ord(char):04x};" for char in chars]
        row = f"\\{i:03o}"[:-1] + "x"
        markdown += f"| **{row}** | {' | '.join(chars)} |\n"
    markdown += "\n"

print(markdown)
```

### CODE REVIEW

File: pygmt/encodings.py
Comment (seisman) [inline pygmt/encodings.py:96]: In previous version, we maintained a big dictionary like:
```
 ✁✂✃✄☎✆✇✈✉☛☞✌✍✎✏
```
but here I decide to use:
```
\u0020\u2701\u2702\u2703\u2704\u260e\u2706\u2707
```
The main reasons are:

- Avoid having too many non-ASCII characters in the source code
- Some non-ASCII characters can't be displayed correctly in text editors or GitHub web UI

File: pygmt/encodings.py
Comment (seisman) [inline pygmt/encodings.py:58]: Or we can just rewrite it like this, which is easier to maintain.
```suggestion
charset["ISO-8859-1"] = dict(
    zip(
        range(0o030, 0o400),
        "\ufffd\u2022\u2026\u2122\u2014\u2013\ufb01\u017e"
        "\u0020\u0021\u0022\u0023\u0024\u0025\u0026\u2019"
        "\u0028\u0029\u002a\u002b\u002c\u2212\u002e\u002f"
        "\u0030\u0031\u0032\u0033\u0034\u0035\u0036\u0037"
        "\u0038\u0039\u003a\u003b\u003c\u003d\u003e\u003f"
        "\u0040\u0041\u0042\u0043\u0044\u0045\u0046\u0047"
        "\u0048\u0049\u004a\u004b\u004c\u004d\u004e\u004f"
        "\u0050\u0051\u0052\u0053\u0054\u0055\u0056\u0057"
        "\u0058\u0059\u005a\u005b\u005c\u005d\u005e\u005f"
        "\u2018\u0061\u0062\u0063\u0064\u0065\u0066\u0067"
        "\u0068\u0069\u006a\u006b\u006c\u006d\u006e\u006f"
        "\u0070\u0071\u0072\u0073\u0074\u0075\u0076\u0077"
        "\u0078\u0079\u007a\u007b\u007c\u007d\u007e\u0161"
        "\u0152\u2020\u2021\u0141\u2044\u2039\u0160\u203a"
        "\u0153\u0178\u017d\u0142\u2030\u201e\u201c\u201d"
        "\u0131\u0060\u00b4\u02c6\u02dc\u00af\u02d8\u02d9"
        "\u00a8\u201a\u02da\u00b8\u0027\u02dd\u02db\u02c7"
        "\u0020\u00a1\u00a2\u00a3\u00a4\u00a5\u00a6\u00a7"
        "\u00a8\u00a9\u00aa\u00ab\u00ac\u002d\u00ae\u00af"
        "\u00b0\u00b1\u00b2\u00b3\u00b4\u00b5\u00b6\u00b7"
        "\u00b8\u00b9\u00ba\u00bb\u00bc\u00bd\u00be\u00bf"
        "\u00c0\u00c1\u00c2\u00c3\u00c4\u00c5\u00c6\u00c7"
        "\u00c8\u00c9\u00ca\u00cb\u00cc\u00cd\u00ce\u00cf"
        "\u00d0\u00d1\u00d2\u00d3\u00d4\u00d5\u00d6\u00d7"
        "\u00d8\u00d9\u00da\u00db\u00dc\u00dd\u00de\u00df"
        "\u00e0\u00e1\u00e2\u00e3\u00e4\u00e5\u00e6\u00e7"
        "\u00e8\u00e9\u00ea\u00eb\u00ec\u00ed\u00ee\u00ef"
        "\u00f0\u00f1\u00f2\u00f3\u00f4\u00f5\u00f6\u00f7"
        "\u00f8\u00f9\u00fa\u00fb\u00fc\u00fd\u00fe\u00ff",
        strict=False,
    )
)
```

File: pygmt/encodings.py
Comment (yvonnefroehlich) [inline pygmt/encodings.py:13]:
```suggestion
Some code points are undefined and are assigned with the replacement character
(``\ufffd``).
```

File: pygmt/encodings.py
Comment (yvonnefroehlich) [inline pygmt/encodings.py:10]:
```suggestion
The corresponding Unicode characters in each Adobe character encoding are generated
from the mapping table and conversion script in the GMT-octal-codes
(https://github.com/seisman/GMT-octal-codes) repository. Refer to that repository for
details.
```

File: doc/techref/index.md
Comment (seisman) [inline doc/techref/index.md:1]: As discussed in https://github.com/GenericMappingTools/pygmt/issues/2834, the documentation group is called "Technical Reference". Here the directory name is `techref`. Please let me know if you're OK with the name.

**Comment by seisman** [comment]
Ping @GenericMappingTools/pygmt-maintainers for reviews and will self-approve and merge in one week if no further comments.

File: doc/techref/index.md
Comment (weiji14) [reply doc/techref/index.md:1]: Yep, ok with the title name. We could maybe add a short paragraph under this title to mention what this page is about? Also would be good to link to https://docs.generic-mapping-tools.org/6.5/reference.html.

File: doc/techref/encodings.md
Comment (weiji14) [inline doc/techref/encodings.md:4]:
```suggestion
characters. Below are some of the most common encodings and characters that are supported.
```

File: doc/techref/encodings.md
Comment (weiji14) [inline doc/techref/encodings.md:7]:
```suggestion
strings. When using non-ASCII characters in PyGMT, the easiest way is to copy and paste
```

File: doc/techref/encodings.md
Comment (weiji14) [inline doc/techref/encodings.md:76]:
```suggestion
**Note**: The octal code `\140` represents the RADICAL EXTENDER character, which is not available in
```

**Review by weiji14 — COMMENTED** [review]
Just some minor typos and one suggestion on adding a short descriptive paragraph in the Technical Reference page. Haven't checked every single character in the tables, but I trust that they're ok :slightly_smiling_face:

File: doc/techref/index.md
Comment (seisman) [reply doc/techref/index.md:1]: Added in f2b3dcb81eebdb402292fef1ad8773ecd905f125.

File: doc/techref/index.md
Comment (weiji14) [inline doc/techref/index.md:6]: Use intersphinx link

```suggestion
visit the :gmt-docs:`GMT Technical Reference <reference.html>`.
```

**Review by weiji14 — APPROVED** [review]

**Pull request merged**
